relay-server/src/actors/spooler/mod.rs

```diff
@@ -501,18 +394,345 @@ impl BufferService {
+            if let Err(key) = self.delete_and_fetch(key, &sender, services).await {
+                unused_keys.insert(key);
+            };
+        }
+        services
+            .project_cache
+            .send(UpdateBufferIndex::new(project_key, unused_keys))
+    }
+
+    /// Returns `true` if the provided used memory can fit on the disk.
+    async fn can_fit(&self, used_memory: usize) -> bool {
+        self.estimate_spool_size()
+            .await
+            .ok()
+            .and_then(|spool_size| self.max_disk_size.checked_sub(spool_size as usize))
+            .map_or(false, |s| s > used_memory)
+    }
+
+    /// Estimates the db size by multiplying `page_count * page_size`.
+    async fn estimate_spool_size(&self) -> Result<i64, BufferError> {
+        let size: i64 = sql::current_size()
+            .fetch_one(&self.db)
+            .await
+            .and_then(|r| r.try_get(0))
+            .map_err(BufferError::FileSizeReadFailed)?;
+
+        relay_statsd::metric!(histogram(RelayHistograms::BufferDiskSize) = size as u64);
+        Ok(size)
+    }
+
+    /// Returns `true` if the maximum size is reached, `false` otherwise.
+    async fn is_full(&self) -> Result<bool, BufferError> {
+        let current_size = self.estimate_spool_size().await?;
+        Ok(current_size as usize >= self.max_disk_size)
+    }
+
+    /// Returns `true` if the spool is empty, `false` otherwise.
+    async fn is_empty(&self) -> Result<bool, BufferError> {
+        let is_empty = sql::select_one()
+            .fetch_optional(&self.db)
+            .await
+            .map_err(BufferError::FetchFailed)?
+            .is_none();
+
+        Ok(is_empty)
+    }
+
+    /// Enqueues data into on-disk spool.
+    async fn enqueue(
+        &self,
+        key: QueueKey,
+        managed_envelope: ManagedEnvelope,
+    ) -> Result<(), BufferError> {
+        let received_at = managed_envelope.received_at().timestamp_millis();
+        sql::insert(
+            key,
+            managed_envelope.into_envelope().to_vec().unwrap(),
+            received_at,
+        )
+        .execute(&self.db)
+        .await
+        .map_err(BufferError::InsertFailed)?;
+
+        relay_statsd::metric!(counter(RelayCounters::BufferWrites) += 1);
+        Ok(())
+    }
+}
+
+/// The state which defines the [`BufferService`] behaviour.
+#[derive(Debug)]
+enum BufferState {
+    /// Only memory buffer is enabled.
+    Memory { ram: InMemory },
+
+    /// The memory buffer is used, but also disk spool is configured.
+    ///
+    /// The disk will be used when the memory limit will be hit.
+    MemoryFileStandby { ram: InMemory, disk: OnDisk },
+
+    /// THe disk is full and we use the memory to save incoming envelopes as much as we can. But
+    /// the disk still used to read from.
+    MemoryFileRead { ram: InMemory, disk: OnDisk },
+
+    /// Only disk used for read/write operations.
+    File { disk: OnDisk },
```

Comment:
nit: I would call this variant `Disk`. The fewer different words, the better :)

```suggestion
    Disk(OnDisk),
```